### PR TITLE
[Placement Group] Allow scheduling a task on any bundle (-1, default)

### LINKED ISSUE #9796: [Placement Group] Should be able to schedule a task on "any bundle"
There is currently no way to allow a task to run on "any bundle". This should be supported if bundle_index = -1.

### PR DESCRIPTION
**Pull request by ericl**
## Why are these changes needed?

This implements wildcard placement for tasks in a placement group, which will be the default behaviour. The task is allowed to be scheduled in any bundle of the group. This is implemented via a pair of custom resources. For example, suppose we placed bundle 3 of placement group ZZZ, and the bundle has two CPU resources. Then, the resources
```
CPU_group_ZZZ
CPU_group_3_ZZZ
```
Will be created on that node. Wildcard tasks will have resource request `{"CPU_group_ZZZ": n}`, while tasks placed in a specific request will have resource request `{"CPU_group_ZZZ": n, "CPU_group_3_ZZZ": n}`.

## Related issue number

Closes https://github.com/ray-project/ray/issues/9796

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: src/ray/common/bundle_spec.cc [hunk 1]
```diff
@@ -66,6 +72,12 @@ std::string FormatPlacementGroupResource(const std::string &original_resource_na
       original_resource_name, bundle_spec.PlacementGroupId(), bundle_spec.Index());
 }
 
+bool IsBundleIndex(const std::string &resource, PlacementGroupID group_id,
+                   const int bundle_index) {
+  return resource.find("_group_" + std::to_string(bundle_index) + "_" + group_id.Hex()) !=
```
Comment (ffbin) [inline src/ray/common/bundle_spec.cc:77]: `"_group_" + std::to_string(bundle_index) + "_" + group_id.Hex()` abstract a function?

File: src/ray/common/bundle_spec.h [hunk 2]
```diff
@@ -99,6 +99,10 @@ std::string FormatPlacementGroupResource(const std::string &original_resource_na
 std::string FormatPlacementGroupResource(const std::string &original_resource_name,
                                          const BundleSpecification &bundle_spec);
 
+/// Return whether a formatted resource is a bundle of the given index.
+bool IsBundleIndex(const std::string &resource, PlacementGroupID group_id,
```
Comment (ffbin) [inline src/ray/common/bundle_spec.h:103]: const PlacementGroupID &group_id?

File: src/ray/common/task/scheduling_resources.cc [hunk 3]
```diff
@@ -227,29 +229,48 @@ void ResourceSet::AddResources(const ResourceSet &other) {
...
+      to_restore[resource_label] = resource_capacity;
       iter = resource_capacity_.erase(iter);
     } else {
       iter++;
     }
   }
+  // For each matching resource to restore (e.g., key like CPU, GPU).
+  for (const auto &pair : to_restore) {
+    resource_capacity_[pair.first] += pair.second;
+    auto wildcard_resource = FormatPlacementGroupResource(pair.first, group_id, -1);
```
Comment (ffbin) [inline src/ray/common/task/scheduling_resources.cc:268]: Add the default parameter -1 to the function declaration of FormatPlacementGroupResource?

File: src/ray/core_worker/core_worker.cc [hunk 4]
```diff
@@ -1173,12 +1173,19 @@ Status CoreWorker::SetResource(const std::string &resource_name, const double ca
 std::unordered_map<std::string, double> AddPlacementGroupConstraint(
     const std::unordered_map<std::string, double> &resources,
     PlacementGroupID placement_group_id, int64_t bundle_index) {
+  if (bundle_index < 0) {
+    RAY_CHECK(bundle_index == -1) << "Invalid bundle index " << bundle_index;
```
Comment (ffbin) [inline src/ray/core_worker/core_worker.cc:1177]: RAY_CHECK(bundle_index >= -1) << "Invalid bundle index " << bundle_index;

**Comment by ffbin** [comment]
Does java code need to be modified synchronously?

File: src/ray/common/bundle_spec.cc [hunk 5]
```diff
@@ -54,8 +54,14 @@ int64_t BundleSpecification::Index() const {
...
                                          int64_t bundle_index) {
-  auto str = original_resource_name + "_group_" + group_id.Hex() + "_" +
-             std::to_string(bundle_index);
+  std::string str;
+  if (bundle_index >= 0) {
+    str = original_resource_name + "_group_" + std::to_string(bundle_index) + "_" +
+          group_id.Hex();
+  } else {
+    RAY_CHECK(bundle_index == -1) << "Invalid index " << bundle_index;
+    str = original_resource_name + "_group_" + group_id.Hex();
```
Comment (ffbin) [inline src/ray/common/bundle_spec.cc:63]: auto str = bundle_index >= 0 ? (original_resource_name + "_group_" + std::to_string(bundle_index) + "_" + group_id.Hex()) : (original_resource_name + "_group_" + group_id.Hex());

File: src/ray/common/bundle_spec.h (hunk 2 above)
Comment (ericl) [reply src/ray/common/bundle_spec.h:103]: Changed.

File: src/ray/common/task/scheduling_resources.cc (hunk 3 above)
Comment (ericl) [reply src/ray/common/task/scheduling_resources.cc:268]: Done.

File: src/ray/core_worker/core_worker.cc (hunk 4 above)
Comment (ericl) [reply src/ray/core_worker/core_worker.cc:1177]: Not needed.

**Comment by ericl** [comment]
> Does java code need to be modified synchronously?

I see some crashing java test, but it seems to be unrelated. Eventually we will need java API parity, but during early development we will prioritize the Python API.

**Comment by ffbin** [comment]
@ericl How about split this issue into two(Java api and Python api) to avoid omission? The java api issue can assign to me.

**Review by rkooo567 — APPROVED** [review]
LGTM!

**Comment by ericl** [comment]
@ffbin sounds great!

**Pull request merged**